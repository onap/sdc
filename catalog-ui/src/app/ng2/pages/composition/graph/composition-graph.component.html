<div class="sdc-composition-graph-wrapper {{zoneTagMode}}"
     [ngClass]="{'with-sidebar': withSidebar$ | async, 'view-only':isViewOnly$ | async}">
</div>

<div class="sdc-composition-menu" [ngClass]="{'with-sidebar': withSidebar$ | async}">

    <service-path-selector
            *ngIf="topologyTemplate.isService() && compositionService.forwardingPaths"
            [drawPath]="drawPathOnCy"
            [deletePaths]="deletePathsOnCy"
            [selectedPathId]="selectedPathId">
    </service-path-selector>

    <canvas-search *ngIf="componentInstanceNames" class="composition-search"
                   [placeholder]="'Type to search'"
                   [data]="componentInstanceNames"
                   (searchChanged)="getAutoCompleteValues($event)"
                   (searchButtonClicked)="highlightSearchMatches($event)">
    </canvas-search>

    <!--<service-path class="zoom-icons"-->
                  <!--*ngIf="!(isViewOnly$ | async) && topologyTemplate.isService()"-->
                  <!--[service]="topologyTemplate"-->
                  <!--[onCreate]="createOrUpdateServicePath">-->
    <!--</service-path>-->

    <svg-icon  *ngIf="!(isViewOnly$ | async) && topologyTemplate.isService()" class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
               [backgroundColor]="'silver'" [name]="'browse'" [clickable]="true" [testId]="'pathsMenuBtn'"
               (click)="openServicePathMenu($event)"></svg-icon>
    <svg-icon class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
              [backgroundColor]="'silver'" [name]="'expand-o'" [clickable]="true"
              (click)="zoomAllWithoutSidebar()"></svg-icon>
    <svg-icon class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
              [backgroundColor]="'silver'" [name]="'plus'" [clickable]="true"
              (click)="zoom(true)"></svg-icon>
    <svg-icon class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
              [backgroundColor]="'silver'" [name]="'minus'" [clickable]="true"
              (click)="zoom(false)"></svg-icon>
</div>

<div class="sdc-canvas-zones__wrapper {{zoneTagMode}}" [ngClass]="{'with-sidebar': withSidebar$ | async}">
    <zone-container *ngFor="let zone of zones" [title]="zone.title" [type]="zone.type" [count]="zone.instances.length"
                    [visible]="zone.visible" [minimized]="zone.minimized" (minimize)="zoneMinimizeToggle(zone.type)"
                    (backgroundClick)="zoneBackgroundClicked()">
        <zone-instance *ngFor="let instance of zone.instances" [hidden]="instance.hidden"
                       [zoneInstance]="instance" [defaultIconText]="zone.defaultIconText"
                       [isActive]="activeZoneInstance == instance"
                       [activeInstanceMode]="activeZoneInstance && activeZoneInstance.mode"
                       [isViewOnly]="isViewOnly$ | async"
                       [forceSave]="instance.forceSave"
                       (modeChange)="zoneInstanceModeChanged($event.newMode, $event.instance, zone.type)"
                       (tagHandleClick)="zoneInstanceTagged($event)"
                       (assignmentSaveStart)="zoneAssignmentSaveStart()"
                       (assignmentSaveComplete)="zoneAssignmentSaveComplete($event)">
        </zone-instance>
    </zone-container>
</div>