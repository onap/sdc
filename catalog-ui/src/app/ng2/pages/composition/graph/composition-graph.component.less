:host(composition-graph) {
  flex: 1;
  padding-top: 53px;
}

.composition {
  .custom-modal {
    /* Hack solution to hide canvas tooltips under modals */
    z-index: 20000 !important;
  }
}

.sdc-composition-graph-wrapper {
  height: 100%;
  width: 100%;

  &.with-sidebar {
    width: calc(~'100% - 300px');
  }
}

.view-only {
  background-color: rgb(248, 248, 248);
}

.sdc-canvas-zones__wrapper {
  position: absolute;
  bottom: 10px;
  right: 12px;
  display: flex;
  transition: right 0.2s;

  &.with-sidebar {
    right: 310px;
  }

  ng2-zone-container {
    display: flex;
    margin-left: 10px;
  }
}

.group-tagging {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/group_1.svg"), pointer;
}

.group-tagging-hover {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/group_2.svg"), pointer;
}

.policy-tagging {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/policy_1.svg"), pointer;
}

.policy-tagging-hover {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/policy_2.svg"), pointer;
}

//Canvas menu
.sdc-composition-menu {
  position: absolute;
  right: 18px;
  top: 53px;
  transition: right 0.2s;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin-right: 10px;
  pointer-events: none;

  & > * {
    pointer-events: all;
  }

  &.with-sidebar {
    right: 320px;
  }

  .composition-search {
    margin-top: 12px;
  }

  .zoom-icons {
    border: solid 1px #d2d2d2;
    border-radius: 2px;
    box-shadow: 0px 2px 3.88px 0.12px rgba(0, 0, 0, 0.29);
    margin-top: 10px;

    /deep/ .svg-icon {
      box-sizing: content-box;
    }
  }
}