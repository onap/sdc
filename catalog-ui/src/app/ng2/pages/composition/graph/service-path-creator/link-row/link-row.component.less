@import './../../../../../../../assets/styles/variables.less';
.remove {
    display: flex;
    align-items: center;
    justify-content: center;
}

.cell {
    padding: 0;
}

/deep/ .link-selector {
    select {
        height: 30px;
        border: none;
        stroke: none;
    }

}


