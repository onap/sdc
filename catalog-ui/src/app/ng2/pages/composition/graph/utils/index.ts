/**
 * Created by ob0695 on 6/3/2018.
 */
// export * from './composition-graph-general-utils';
// export * from './composition-graph-links-utils';
// export * from './composition-graph-nodes-utils';
// export * from './composition-graph-palette-utils';
// export * from './composition-graph-service-path-utils';
// export * from './composition-graph-zone-utils';


import {CompositionGraphGeneralUtils} from './composition-graph-general-utils';
import {CompositionGraphNodesUtils} from './composition-graph-nodes-utils';
import {MatchCapabilitiesRequirementsUtils} from './match-capability-requirement-utils'
import {CompositionGraphPaletteUtils} from './composition-graph-palette-utils';
import {CompositionGraphZoneUtils} from './composition-graph-zone-utils';
import {ServicePathGraphUtils} from './composition-graph-service-path-utils';
import {CompositionGraphLinkUtils} from "./composition-graph-links-utils";


export {
    CompositionGraphGeneralUtils,
    CompositionGraphLinkUtils,
    CompositionGraphNodesUtils,
    MatchCapabilitiesRequirementsUtils,
    CompositionGraphPaletteUtils,
    CompositionGraphZoneUtils,
    ServicePathGraphUtils
};