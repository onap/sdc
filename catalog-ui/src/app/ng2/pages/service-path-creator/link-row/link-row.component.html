<!--
  ~ Copyright (C) 2018 AT&T Intellectual Property. All rights reserved.
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~      http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->
 
<ui-element-dropdown
	data-tests-id="linkSrc"
	[readonly]="!link.isFirst || (link.isFirst && !link.canEdit)"
	class="cell link-selector"
	[values]="source"
	[(value)]="link.fromNode"
	(valueChange)="onSourceSelected($event)">
</ui-element-dropdown>

<ui-element-dropdown
	data-tests-id="linkSrcCP"
	[readonly]="!link.isFirst || (link.isFirst && !link.canEdit)"
	class="cell link-selector"
	[values]="srcCP"
	[(value)]="link.fromCP"
	(valueChange)="onSrcCPSelected($event)">
</ui-element-dropdown>

<ui-element-dropdown
	data-tests-id="linkTarget"
	[readonly]="!link.canEdit"
	class="cell link-selector"
	[values]="target"
	[(value)]="link.toNode"
	(valueChange)="onTargetSelected($event)">
</ui-element-dropdown>

<ui-element-dropdown
	data-tests-id="linkTargetCP"
	[readonly]="!link.canEdit"
	class="cell link-selector"
	[values]="targetCP"
	[(value)]="link.toCP"
	(valueChange)="onTargetCPSelected($event)">
</ui-element-dropdown>

<div
	class="cell remove"
	data-tests-id="removeLnk">
	<span
		*ngIf="link.canRemove"
		class="sprite-new delete-item-icon"
		(click)="removeRow()">
	</span>
</div>
