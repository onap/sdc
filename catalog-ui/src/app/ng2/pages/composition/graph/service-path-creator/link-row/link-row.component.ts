import {Component, Input} from '@angular/core';
import {DropdownValue} from "app/ng2/components/ui/form-components/dropdown/ui-element-dropdown.component";
import {Link} from './link.model';
import {ServicePathMapItem} from "app/models/graph/nodes-and-links-map";
import * as _ from "lodash";

@Component({
    selector: 'link-row',
    templateUrl: './link-row.component.html',
    styleUrls: ['./link-row.component.less']
})


export class LinkRowComponent {
    @Input() data:Array<ServicePathMapItem>;
    @Input() link:Link;
    @Input() removeRow:Function;
    source: Array<DropdownValue> = [];
    target: Array<DropdownValue> = [];
    srcCP: Array<DropdownValue> = [];
    targetCP: Array<DropdownValue> = [];

    ngOnChanges() {
        if (this.data) {
            this.parseInitialData(this.data);
        }
    }

    parseInitialData(data: Array<ServicePathMapItem>) {
        this.source = this.convertValuesToDropDownOptions(data);
        if (this.link.fromNode) {
            let srcCPOptions = this.findOptions(data, this.link.fromNode);
            if (!srcCPOptions) { return; }
            this.srcCP = this.convertValuesToDropDownOptions(srcCPOptions);
            if (this.link.fromCP) {
                this.target = this.convertValuesToDropDownOptions(data);
                if (this.link.toNode) {
                    let targetCPOptions = this.findOptions(data, this.link.toNode);
                    if (!targetCPOptions) { return; }
                    this.targetCP = this.convertValuesToDropDownOptions(targetCPOptions);
                }
            }
        }
    }

    private findOptions(items: Array<ServicePathMapItem>, nodeOrCPId: string) {
        let item = _.find(items, (dataItem) => nodeOrCPId === dataItem.id);
        if (item && item.data && item.data.options) {
            return item.data.options;
        }
        console.warn('no option was found to match selection of Node/CP with id:' + nodeOrCPId);
        return null;
    }

    private convertValuesToDropDownOptions(values: Array<ServicePathMapItem>): Array<DropdownValue> {
        let result:Array<DropdownValue> = [];
        for (let i = 0; i < values.length ; i++) {
            result[result.length] =  new DropdownValue(values[i].id, values[i].data.name);
        }
        return result.sort((a, b) => a.label.localeCompare(b.label));
    }

    onSourceSelected(id) {
        if (id) {
            let srcCPOptions = this.findOptions(this.data, id);
            this.srcCP = this.convertValuesToDropDownOptions(srcCPOptions);
            this.link.fromCP = '';
            this.link.toNode = '';
            this.link.toCP = '';
            this.target = [];
            this.targetCP = [];
        }
    }

    onSrcCPSelected (id) {
        if (id) {
            let srcCPOptions = this.findOptions(this.data, this.link.fromNode);
            let srcCPData = srcCPOptions.find(option => id === option.id).data;
            this.target = this.convertValuesToDropDownOptions(this.data);
            this.link.fromCPOriginId = srcCPData.ownerId;
            this.link.toNode = '';
            this.link.toCP = '';
            this.targetCP = [];
        }

    }

    onTargetSelected(id) {
        if (id) {
            let targetCPOptions = this.findOptions(this.data, id);
            this.targetCP = this.convertValuesToDropDownOptions(targetCPOptions);
            this.link.toCP = '';
        }

    }

    onTargetCPSelected(id) {
        if (id) {
            let targetCPOptions = this.findOptions(this.data, this.link.toNode);
            let targetCPDataObj = targetCPOptions.find(option => id === option.id).data;
            this.link.toCPOriginId = targetCPDataObj.ownerId;
        }
    }
}
