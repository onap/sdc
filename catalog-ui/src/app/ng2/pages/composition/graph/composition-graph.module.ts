import {NgModule} from "@angular/core";
import {CommonModule} from "@angular/common";
import {CompositionGraphComponent} from "./composition-graph.component";
import {ZoneModules} from "./canvas-zone/zones-module";
import {CompositionGraphZoneUtils} from "./utils/composition-graph-zone-utils";
import {CompositionGraphGeneralUtils} from "./utils/composition-graph-general-utils";
import {CommonGraphUtils} from "./common/common-graph-utils";
import {LinksFactory} from "app/models/graph/graph-links/links-factory";
import {NodesFactory} from "app/models/graph/nodes/nodes-factory";
import {ImageCreatorService} from "./common/image-creator.service";
import {MatchCapabilitiesRequirementsUtils} from "./utils/match-capability-requierment-utils";
import {CompositionGraphNodesUtils} from "./utils/composition-graph-nodes-utils";
import {ConnectionWizardService} from "app/ng2/pages/composition/graph/connection-wizard/connection-wizard.service";
import {CompositionGraphPaletteUtils} from "./utils/composition-graph-palette-utils";
import {QueueServiceUtils} from "app/ng2/utils/queue-service-utils";
import {DndModule} from "ngx-drag-drop";
import { MenuListNg2Module } from "app/ng2/components/downgrade-wrappers/menu-list-ng2/menu-list-ng2.module";
import { UiElementsModule } from "app/ng2/components/ui/ui-elements.module";
import {ServicePathSelectorModule} from "./service-path-selector/service-path-selector.module";
import {SdcUiComponentsModule, SdcUiServices} from "onap-ui-angular";
import {CanvasSearchModule} from "./canvas-search/canvas-search.module";
import {CompositionGraphLinkUtils, ServicePathGraphUtils} from "./utils";


@NgModule({
    declarations: [CompositionGraphComponent],
    imports: [CommonModule,
        ServicePathSelectorModule,
        SdcUiComponentsModule,
        MenuListNg2Module,
        UiElementsModule,
        ZoneModules,
        CanvasSearchModule,
        DndModule],
    exports: [CompositionGraphComponent],
    entryComponents: [CompositionGraphComponent],
    providers: [
        CompositionGraphZoneUtils,
        CompositionGraphGeneralUtils,
        MatchCapabilitiesRequirementsUtils,
        CompositionGraphNodesUtils,
        CompositionGraphLinkUtils,
        CommonGraphUtils,
        NodesFactory,
        LinksFactory,
        ImageCreatorService,
        ConnectionWizardService,
        CompositionGraphPaletteUtils,
        QueueServiceUtils,
        SdcUiServices.simplePopupMenuService,
        ServicePathGraphUtils
    ]
})
export class CompositionGraphModule {
}