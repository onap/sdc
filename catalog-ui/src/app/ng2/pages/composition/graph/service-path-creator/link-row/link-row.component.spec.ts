import {async, ComponentFixture} from "@angular/core/testing";
import {CacheService} from "../../../../../services/cache.service";
import {ConfigureFn, configureTests} from "../../../../../../../jest/test-config.helper";
import {NO_ERRORS_SCHEMA} from "@angular/core";
import {LinkRowComponent} from "./link-row.component";
import {DropdownValue} from "../../../../../components/ui/form-components/dropdown/ui-element-dropdown.component";
import {MapItemData, ServicePathMapItem} from "../../../../../../models/graph/nodes-and-links-map";

describe('artifact form component', () => {

    let fixture: ComponentFixture<LinkRowComponent>;
    let cacheServiceMock: Partial<CacheService>;

    beforeEach(
        async(() => {


            cacheServiceMock = {
                contains: jest.fn(),
                remove: jest.fn(),
                set: jest.fn(),
                get: jest.fn()
            }

            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [LinkRowComponent],
                    imports: [],
                    schemas: [NO_ERRORS_SCHEMA],
                    providers: []
                    ,
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(LinkRowComponent);
            });
        })
    );


    it('should match current snapshot of artifact form component', () => {
        expect(fixture).toMatchSnapshot();
    });


    it('ngOnChanges() -> in case data exist -> call to parseInitialData()' ,() => {
        // init values / mock functions
        let data = 'something';
        fixture.componentInstance.parseInitialData = jest.fn();
        fixture.componentInstance.data = data;

        // call to the tested function
        fixture.componentInstance.ngOnChanges();

        // expect that
        expect(fixture.componentInstance.parseInitialData).toHaveBeenCalledWith(data);
    });

    it('onSourceSelected() -> in case id -> srcCP, link.fromCP, link.toNode, link.toCP, target, targetCP should be updated accordingly' ,() => {
        // init values / mock functions
        let id = 'id';
        let data = 'data';
        let link = {
            fromCP:'testVal',
            toNode:'testVal',
            toCP:'testVal'
        }
        let target = ['val1', 'val2'];
        let targetCP = ['val1', 'val2'];

        fixture.componentInstance.findOptions = jest.fn();
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => 'dummyConvertedVal');
        fixture.componentInstance.data = data;
        fixture.componentInstance.link = link;
        fixture.componentInstance.target = target;
        fixture.componentInstance.targetCP = targetCP;

        // call to the tested function
        fixture.componentInstance.onSourceSelected(id);

        // expect that
        expect(fixture.componentInstance.findOptions).toHaveBeenCalledWith(data, id);
        expect(fixture.componentInstance.srcCP).toBe('dummyConvertedVal');
        expect(fixture.componentInstance.link.fromCP).toBe('');
        expect(fixture.componentInstance.link.toNode).toBe('');
        expect(fixture.componentInstance.link.toCP).toBe('');
        expect(fixture.componentInstance.target.length).toBe(0);
        expect(fixture.componentInstance.targetCP.length).toBe(0);
    });

    it('onSourceSelected() -> in case id undefined -> No Change to srcCP, link.fromCP, link.toNode, link.toCP, target, targetCP' ,() => {
        // init values / mock functions
        let id;
        let data = 'data';
        let link = {
            fromCP:'testVal',
            toNode:'testVal',
            toCP:'testVal'
        }
        let target = ['val1', 'val2'];
        let targetCP = ['val1', 'val2'];

        fixture.componentInstance.findOptions = jest.fn();
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => 'dummyConvertedVal');
        fixture.componentInstance.data = data;
        fixture.componentInstance.link = link;
        fixture.componentInstance.target = target;
        fixture.componentInstance.targetCP = targetCP;

        // call to the tested function
        fixture.componentInstance.onSourceSelected(id);

        // expect that
        expect(fixture.componentInstance.link.fromCP).toBe(link.fromCP);
        expect(fixture.componentInstance.link.toNode).toBe(link.toNode);
        expect(fixture.componentInstance.link.toCP).toBe(link.toCP);
        expect(fixture.componentInstance.target.length).toBe(2);
        expect(fixture.componentInstance.target[0]).toBe('val1')
        expect(fixture.componentInstance.targetCP.length).toBe(2);
        expect(fixture.componentInstance.targetCP[1]).toBe('val2');
    });

    it('onSrcCPSelected() -> in case id  -> Verify target, link.fromCPOriginId, link.toNode, link.toCP, targetCP.length' ,() => {
        // init values / mock functions
        let id = 'id';
        let link = {
            fromNode:'testVal',
            toCPOriginId: 'initValue_ShouldBeChanged'
        };
        let option1 = {
            id: 'something'
        };
        let option2 = {
            id: 'id',
            data: {"ownerId":1}
        };

        fixture.componentInstance.link = link;
        fixture.componentInstance.findOptions = jest.fn(() => [option1, option2]);
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => 'dummyConvertedVal');

        // call to the tested function
        fixture.componentInstance.onSrcCPSelected(id);

        // expect that
        expect(fixture.componentInstance.target).toBe('dummyConvertedVal');
        expect(fixture.componentInstance.link.fromCPOriginId).toBe(option2.data.ownerId);
        expect(fixture.componentInstance.link.toNode).toBe('');
        expect(fixture.componentInstance.link.toCP).toBe('');
        expect(fixture.componentInstance.targetCP.length).toBe(0);

    });

    it('onSrcCPSelected() -> in case id undefined -> Verify target, link.fromCPOriginId, link.toNode, link.toCP, targetCP.length' ,() => {
        // init values / mock functions
        let id;

        let targetInput:Array<DropdownValue> = [{value:'Value', label:'Label', hidden:true, selected:true}];

        let linkInput = {
            fromCPOriginId:'expectedLinkFromCPOriginId',
            toNode:'expectedLinkToNode',
            toCP:'expectedLinkToCP',
            // Link Object
            canEdit:true,
            canRemove:true,
            isFirst:true,
            // ForwardingPathLink Object
            ownerId:'',
            fromNode:'',
            fromCP:'',
            toCPOriginId:''
        }

        fixture.componentInstance.target = targetInput;
        fixture.componentInstance.link = linkInput;
        fixture.componentInstance.targetCP = targetInput;


        // call to the tested function
        fixture.componentInstance.onSrcCPSelected(id);

        // expect that
        expect(fixture.componentInstance.target).toBe(targetInput);
        expect(fixture.componentInstance.link.fromCPOriginId).toBe('expectedLinkFromCPOriginId');
        expect(fixture.componentInstance.link.toNode).toBe('expectedLinkToNode');
        expect(fixture.componentInstance.link.toCP).toBe('expectedLinkToCP');
        expect(fixture.componentInstance.targetCP.length).toBe(1);
    });

    it('onTargetSelected() -> in case id  -> Verify targetCP & link.toCP' ,() => {
        // init values / mock functions
        let id = 'id';
        let link = {
            toCP:'testVal'
        }
        let targetCP = ['val1', 'val2'];

        fixture.componentInstance.findOptions = jest.fn();
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => 'dummyConvertedVal');
        fixture.componentInstance.link = link;
        fixture.componentInstance.targetCP = targetCP;

        // call to the tested function
        fixture.componentInstance.onTargetSelected(id);

        // expect that
        expect(fixture.componentInstance.targetCP).toBe('dummyConvertedVal');
        expect(fixture.componentInstance.link.toCP).toBe('');

    });

    it('onTargetSelected() -> in case id undefined -> Verify targetCP & link.toCP' ,() => {
        // init values / mock functions
        let id;
        let link = {
            toCP:'toCP_testVal'
        }
        let targetCP = ['val1', 'val2'];

        fixture.componentInstance.findOptions = jest.fn();
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => 'dummyConvertedVal');
        fixture.componentInstance.link = link;
        fixture.componentInstance.targetCP = targetCP;

        // call to the tested function
        fixture.componentInstance.onTargetSelected(id);

        // expect that
        expect(fixture.componentInstance.targetCP.length).toBe(2);
        expect(fixture.componentInstance.targetCP).toEqual(['val1', 'val2']);
        expect(fixture.componentInstance.link.toCP).toBe('toCP_testVal');
    });

    it('onTargetCPSelected() -> in case id  -> Validate toCPOriginId' ,() => {
        // init values / mock functions
        let id = 'id';
        let link = {
            toNode:'testVal',
            toCPOriginId: 'initValue_ShouldBeChanged'
        };
        let option1 = {
             id: 'something'
        };
        let option2 = {
             id: 'id',
             data: {"ownerId":1}
        };
        fixture.componentInstance.link = link;
        fixture.componentInstance.findOptions = jest.fn(() => [option1, option2]);

        // call to the tested function
        fixture.componentInstance.onTargetCPSelected(id);

        // expect that
        expect(fixture.componentInstance.link.toCPOriginId).toBe(option2.data.ownerId);
    });

    it('onTargetCPSelected() -> in case id undefined -> Validate toCPOriginId' ,() => {
        // init values / mock functions
        let id;
        let link = {
            toNode:'testVal',
            toCPOriginId: 'initValue_ShouldRemain'
        };
        let option1 = {
            id: 'something'
        };
        let option2 = {
            id: 'id',
            data: {"ownerId":1}
        };
        fixture.componentInstance.link = link;
        fixture.componentInstance.findOptions = jest.fn(() => [option1, option2]);

        // call to the tested function
        fixture.componentInstance.onTargetCPSelected(id);

        // expect that
        expect(fixture.componentInstance.link.toCPOriginId).toBe('initValue_ShouldRemain');
    });


    it('findOptions() -> in case item.data.options -> Validate return item.data.options' ,() => {
        // init values / mock functions
        const innerMapItemData1: MapItemData = { id: 'innerMapItemData1_id', name: 'innerMapItemData1_name', options: []};
        const innerServicePathItem: ServicePathMapItem = { id: 'innerServicePathItem_id', data: innerMapItemData1 };
        const mapItemData1: MapItemData = { id: 'mapItemData1_id', name: 'mapItemData1_name', options: [innerServicePathItem]};

        const servicePathItem: ServicePathMapItem = { id: 'servicePathItem_id', data: mapItemData1 };
        const arrServicePathItems: ServicePathMapItem[] = [servicePathItem];

        let nodeOrCPId: string = servicePathItem.id;

        // call to the tested function
        let res = fixture.componentInstance.findOptions(arrServicePathItems, nodeOrCPId);

        // expect that
        expect(res).toEqual([innerServicePathItem]);
    });

    it('findOptions() -> in case NOT item || item.data || item.data.options -> Validate return null' ,() => {
        // init values / mock functions
        let item = [{
            // data: {
                data:{
                    name:'data_name',
                    id: 'data_id'
                },
                name:'name',
                id: 'id'
            // }
        }];
        let items: Array<ServicePathMapItem> = item;
        let nodeOrCPId: string = 'someString';

        // call to the tested function
        let res = fixture.componentInstance.findOptions(items, nodeOrCPId);

        // expect that
        expect(res).toBe(null);
    });

    it('convertValuesToDropDownOptions() -> Verify that the result is sorted' ,() => {
        // init values / mock functions
        const mapItemData1: MapItemData = { id: 'Z_ID', name: 'Z_NAME'};
        const servicePathItem1: ServicePathMapItem = { id: 'Z_servicePathItem_id', data: mapItemData1 };

        const mapItemData2: MapItemData = { id: 'A_ID', name: 'A_NAME'};
        const servicePathItem2: ServicePathMapItem = { id: 'A_servicePathItem_id', data: mapItemData2 };

        const mapItemData3: MapItemData = { id: 'M_ID', name: 'M_NAME'};
        const servicePathItem3: ServicePathMapItem = { id: 'M_servicePathItem_id', data: mapItemData3 };

        const arrServicePathItems: ServicePathMapItem[] = [servicePathItem1, servicePathItem2, servicePathItem3];

        // call to the tested function
        let res = fixture.componentInstance.convertValuesToDropDownOptions(arrServicePathItems);

        // expect that
        expect(res.length).toBe(3);
        expect(res[0].value).toBe("A_servicePathItem_id");
        expect(res[0].label).toBe("A_NAME");
        expect(res[1].value).toBe("M_servicePathItem_id");
        expect(res[1].label).toBe("M_NAME");
        expect(res[2].value).toBe("Z_servicePathItem_id");
        expect(res[2].label).toBe("Z_NAME");

    });

    it('parseInitialData() -> link.fromNode Exist => Verify srcCP' ,() => {
        // init values / mock functions

        //Simulate Array<ServicePathMapItem to pass to the function
        const mapItemData1: MapItemData = { id: 'mapItemID', name: 'mapItemName'};
        const servicePathItem1: ServicePathMapItem = { id: 'servicePathItemId', data: mapItemData1 };
        const arrServicePathItems: ServicePathMapItem[] = [servicePathItem1];

        //Simulate link
        let link = {
            fromNode:'testVal'
        };
        fixture.componentInstance.link = link;

        //Simulate the response from convertValuesToDropDownOptions()
        const value = "expected_id_fromNode";
        const label = "expected_label_fromNode"
        let result:Array<DropdownValue> = [];
        result[0] =  new DropdownValue(value, label);
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => result);

        //Simulate the response from findOptions()
        const innerMapItemData1: MapItemData = { id: 'innerMapItemData1_id', name: 'innerMapItemData1_name', options: []};
        const options: ServicePathMapItem = { id: 'innerServicePathItem_id', data: innerMapItemData1 };
        fixture.componentInstance.findOptions = jest.fn(() => options);


        // call to the tested function
        fixture.componentInstance.parseInitialData(arrServicePathItems);

        // expect that
        expect(fixture.componentInstance.srcCP.length).toBe(1);
        expect(fixture.componentInstance.srcCP[0]).toEqual({
            "value": value,
            "label": label,
            "hidden": false,
            "selected": false
        });
    });

    it('parseInitialData() -> link.fromNode & link.fromCP Exist => Verify srcCP' ,() => {
        // init values / mock functions

        //Simulate Array<ServicePathMapItem to pass to the function
        const mapItemData1: MapItemData = { id: 'mapItemID', name: 'mapItemName'};
        const servicePathItem1: ServicePathMapItem = { id: 'servicePathItemId', data: mapItemData1 };
        const arrServicePathItems: ServicePathMapItem[] = [servicePathItem1];

        //Simulate link
        let link = {
            fromNode:'testVal',
            fromCP: 'testVal'
        };
        fixture.componentInstance.link = link;

        //Simulate the response from convertValuesToDropDownOptions()
        const value = "expected_id_fromNode_and_fromCP";
        const label = "expected_label_fromNode_and_fromCP"
        let result:Array<DropdownValue> = [];
        result[0] =  new DropdownValue(value, label);
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => result);

        //Simulate the response from findOptions()
        const innerMapItemData1: MapItemData = { id: 'innerMapItemData1_id', name: 'innerMapItemData1_name', options: []};
        const options: ServicePathMapItem = { id: 'innerServicePathItem_id', data: innerMapItemData1 };
        fixture.componentInstance.findOptions = jest.fn(() => options);


        // call to the tested function
        fixture.componentInstance.parseInitialData(arrServicePathItems);

        // expect that
        expect(fixture.componentInstance.srcCP.length).toBe(1);
        expect(fixture.componentInstance.srcCP[0]).toEqual({
            "value": value,
            "label": label,
            "hidden": false,
            "selected": false
        });
    });


    it('parseInitialData() -> link.fromNode & link.fromCP & link.toNode Exist => Verify srcCP' ,() => {
        // init values / mock functions

        //Simulate Array<ServicePathMapItem to pass to the function
        const mapItemData1: MapItemData = { id: 'mapItemID', name: 'mapItemName'};
        const servicePathItem1: ServicePathMapItem = { id: 'servicePathItemId', data: mapItemData1 };
        const arrServicePathItems: ServicePathMapItem[] = [servicePathItem1];

        //Simulate link
        let link = {
            fromNode:'testVal',
            fromCP: 'testVal',
            toNode: 'testVal'
        };
        fixture.componentInstance.link = link;

        //Simulate the response from convertValuesToDropDownOptions()
        const value = "expected_id_fromNode_and_fromCP_and_toNode";
        const label = "expected_label_fromNode_and_fromCP_and_toNode"
        let result:Array<DropdownValue> = [];
        result[0] =  new DropdownValue(value, label);
        fixture.componentInstance.convertValuesToDropDownOptions = jest.fn(() => result);

        //Simulate the response from findOptions()
        const innerMapItemData1: MapItemData = { id: 'innerMapItemData1_id', name: 'innerMapItemData1_name', options: []};
        const options: ServicePathMapItem = { id: 'innerServicePathItem_id', data: innerMapItemData1 };
        fixture.componentInstance.findOptions = jest.fn(() => options);


        // call to the tested function
        fixture.componentInstance.parseInitialData(arrServicePathItems);

        // expect that
        expect(fixture.componentInstance.srcCP.length).toBe(1);
        expect(fixture.componentInstance.srcCP[0]).toEqual({
            "value": value,
            "label": label,
            "hidden": false,
            "selected": false
        });
    });



});