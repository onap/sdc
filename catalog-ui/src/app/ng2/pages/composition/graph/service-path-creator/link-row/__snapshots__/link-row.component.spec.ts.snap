// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`artifact form component should match current snapshot of artifact form component 1`] = `
<link-row
  source={[Function Array]}
  srcCP={[Function Array]}
  target={[Function Array]}
  targetCP={[Function Array]}
>
  <ui-element-dropdown
    class="cell link-selector"
    data-tests-id="linkSrc"
  /><ui-element-dropdown
    class="cell link-selector"
    data-tests-id="linkSrcCP"
  /><ui-element-dropdown
    class="cell link-selector"
    data-tests-id="linkTarget"
  /><ui-element-dropdown
    class="cell link-selector"
    data-tests-id="linkTargetCP"
  /><div
    class="cell remove"
    data-tests-id="removeLnk"
  >
    
  </div>
</link-row>
`;
